### PR TITLE
Where is your data?

### PR DESCRIPTION
**Pull request by weeman1337**
For previous discussions also see #11319.

**ToDo**
- [x] personal settings layout improvement
- [x] settings input view

### CODE REVIEW

**Comment by weeman1337** [comment]
That's what the front end looks so far:
image

For me it's hard to distinguish links and text as they're displayed the same. So I highlighted them with the primary color. I didn't find anything about that in the dev docs.

The "... is your admin" will have two ways for detection:
- Any admin user if none was selected in the settings
- The selected one from the settings

Any thoughts about that @nextcloud/designers ?

**Comment by skjnldsv** [comment]
I really like it!
I agree with Jan's suggestions. I would also add that the plug icon doesn't really make any sense. Maybe use a [server icon](https://thenounproject.com/search/?q=server)?

**Comment by weeman1337** [comment]
> I would also add that the plug icon doesn't really make any sense

Fully agree. This is only a placeholder - I'm not happy with it, too.  
Should I extend the [NC icon set](https://docs.nextcloud.com/server/15/developer_manual/design/icons.html)? Is there any free icon set I can take the icon from?

> We could think about using "border-bottom: 1px dotted text-color" for marking links.

✓ will do that

> For admins, it's simpler to use "… contact them." instead of "him/her". Works as singular and plural too if there's multiple admins.

@jancborchardt do you suggest listing all admin users then? I thought about picking one of the admins and in addition give the option to define a specific one in the settings.

**Comment by skjnldsv** [comment]
> @jancborchardt do you suggest listing all admin users then? I thought about picking one of the admins and in addition give the option to define a specific one in the settings.

That is one issue we have with the preferred provider program.
There is no official setting to define a 'contact' for a nextcloud instance. It could be a nice addition :)

**Comment by shironextcloud** [comment]
>     * For admins, it's simpler to use "… contact them." instead of "him/her". Works as singular and plural too if there's multiple admins.
> 
>     * @shironextcloud what do you think? :)

agree on "them"

really nice :)

**Comment by weeman1337** [comment]
@nextcloud/designers regarding links, here are the two dotted options:

text-decoration
---
border-bottom
---
Where the `text-decoration` way looks cleaner on the whole `border-bottom` is easier to read.

**Comment by weeman1337** [comment]
Here is the view for the settings:

image

I placed it in the "Basic settings" at the bottom.

**Comment by skjnldsv** [comment]
It looks amazing!!!
Can I be peaky and ask the loader to be on the button to save the form? :see_no_evil: We tend to do that in most of the places and looks better imho. Not that big of a deal though ;)

**Comment by weeman1337** [comment]
> Can I be peaky and ask the loader to be on the button to save the form? see_no_evil We tend to do that in most of the places and looks better imho. Not that big of a deal though ;)

@skjnldsv jep - fine for me. Can you give me an example, so I can have a look copy the style.

There is also some feedback (I took it from the settings above). Does anything common exist for that, too?

image

**Comment by skjnldsv** [comment]
Could be nice to have a standard actually!
You know what, I'll let you decide! We could then make sure to have the new standard in as many places as we can on the settings! :rocket: :bowing_man:

**Comment by weeman1337** [comment]
@skjnldsv what about this:
button

**Comment by skjnldsv** [comment]
I really like it!! :tada: :heart_eyes: 
Some quick thinking:
- I would actually make sure the size does not change that much, because it's confusing to the user.
- I would also not use an input since the loader cannot work (fallback to gif since no pseudo-elements can be applied on an inline element) Use a label to link the action to the input! :wink: 
- No need for the `...` since the loader already give a progress state
- I would use an icon by default for the save button. It draws attention and would also prevent your button from moving too much when switching from "save" to "saving" since the loader width would already be taken by the icon 

I'm sure @jancborchardt  will have some insights as well!

**Comment by weeman1337** [comment]
Update today: the server info is now stored and loaded.

I put the info into the config. Why?: My idea is if you deploy an instance you can simply put it into the file. If anyone thinks it should be a db setting, let me know.

The funky save button also got an update: The spinner shows with 1s delay to prevent unnecessary flickering. It says "saved" until you change any form value. Works fine so far. With another icon for the default state ("save") it doesn't jump that much.

**Comment by jancborchardt** [comment]
Agree with what @skjnldsv said, some additional feedback:

- Buttons should usually go on the right, but being aligned with the fields seems good too, and the size switching will be less strange since the icon and the start of the text will always stay
- The Save button should have `class="primary"` to appear blue
- "Saving" and "Saved" states should have a capital first letter
- The placeholders of the input fields should also have a capital first letter
- For the "Website" and "Link to privacy policy" fields, the placeholder should be "https://…"
- Actually thinking about it: Shouldn’t the company/instance name, website link and privacy policy link be taken from Theming, where we already have that info? cc @juliushaertl 
- Let’s use border-bottom for the links for now. We have to do this generally in the server but this is a good start. (Please make sure to prefix it so it only changes the link styles in this specific container)

If there is no data put in for one of the parts yet, like the server location for example, that part should simply not be shown in the list.

**Comment by jancborchardt** [comment]
Sorry, forgot to answer your question. For the provider icon, simply use `icon-home`:

https://docs.nextcloud.com/server/15/developer_manual/design/icons.html

**Comment by weeman1337** [comment]
> Actually thinking about it: Shouldn’t the company/instance name, website link and privacy policy link be taken from Theming, where we already have that info?

Yes, I see some of the info is already there. Having it redundant doesn't make sense. I think "name" in theming is more the instance name. So missing is "server location", "provider name" and "admin contact". I could also add these fields to the theming section. @nextcloud/designers  ¯\\\_(ツ)\_/¯

**Comment by weeman1337** [comment]
Ping @juliushaertl - waiting for a response where to put the fields..

**Comment by skjnldsv** [comment]
I think there should be a dedicated option to it. Theming is design only, nothing related to administration or mails should go there I think :)

**Comment by jancborchardt** [comment]
> Theming is design only, nothing related to administration or mails should go there I think :)

That’s not true though. ;) This is the one place where you set things like the instance name, link, privacy policy & terms links etc. These are used across the app, including in emails and such. It’s good that it’s only this one place, and hence the things we need for "Who Owns Your Data" should also either be pulled from there, or if it extends it, be added there.

@weeman1337 the Unsplash app also adds a section in the Theming settings, maybe you can check that out?

**Comment by MariusBluem** [comment]
„Your Data is protected by“ should be changed to „Your Files are encrypted with“ since

- encryption is only about files and not about all data (maybe we should clarify this even more)
- encrypted is more neutral than protect

**Comment by skjnldsv** [comment]
> That’s not true though. ;) This is the one place where you set things like the instance name, link, privacy policy & terms links etc. These are used across the app, including in emails and such. It’s good that it’s only this one place, and hence the things we need for "Who Owns Your Data" should also either be pulled from there, or if it extends it, be added there.

Oh you're right! I forgot we added this!
Theming doesn't seems like the appropriate word now! :)
Then yes, This should go into theming I think!

**Comment by juliushaertl** [comment]
Sorry, missed the notification about this one. Yes, a separate section inside of the theming settings seems suitable to me.

**Comment by MorrisJobke** [comment]
> Sorry, missed the notification about this one. Yes, a separate section inside of the theming settings seems suitable to me.

Do we really want to bundle this with the theming app, because then it can't be used without the theming app?

**Comment by juliushaertl** [comment]
> Do we really want to bundle this with the theming app, because then it can't be used without the theming app?

The idea was more to just use the Theming section in the settings, it should still be separate from the theming app itself.

**Comment by weeman1337** [comment]
@MariusBluem I've updated the wording:
image

@juliushaertl the server info is now in the theming section. That means I made the theming section a built in one.

**Comment by MariusBluem** [comment]
> @MariusBluem I've updated the wording:

Thanks 👍

What do you think about "About this server" instead of "Where is your data?"? @jancborchardt

**Comment by jancborchardt** [comment]
@MariusBluem I’d say let’s do any further changes in follow-up pull requests. :) The "Where is your data?" is direct and nice.

cc @nextcloud/javascript @nextcloud/designers for review as well :slightly_smiling_face:

**Review by MorrisJobke — COMMENTED** [review]
Tested and works 👍 

The only minor part that would be nice to be fixed is that if there is nothing specified also the header should not be shown:

bildschirmfoto 2019-01-28 um 16 45 11

**Review by skjnldsv — CHANGES_REQUESTED** [review]
See Morris's comment :)

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

File: settings/templates/settings/personal/partials/where-is-your-data.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		</div>
+	<?php endif; ?>
+
+	<?php if (empty($_['adminName']) === false): ?>
+	<div class="personal-info icon-user-admin">
+		<p>
+			<?php echo $l->t(
+				'%s%s%s is your admin. If you have any issues, %scontact them%s.',
+				[
+					'<a href="mailto:' . $_['adminMail'] . '" target="_blank" title="" rel="noreferrer noopener">',
```
Comment (danielkesselberg) [inline settings/templates/settings/personal/partials/where-is-your-data.php]: In theory there could be an admin account without email. I guess we could ignore that for now.

File: lib/private/Settings/Personal/PersonalInfo.php [hunk 2]
```diff
@@ -71,7 +75,8 @@ public function __construct(
 		AccountManager $accountManager,
 		IAppManager $appManager,
 		IFactory $l10nFactory,
-		IL10N $l
+		IL10N $l,
+		EncryptionManager $encryptionManager
```
Comment (danielkesselberg) [inline lib/private/Settings/Personal/PersonalInfo.php]: PHPDoc missing for $encryptionManager

**Review by kesselb — CHANGES_REQUESTED** [review]
:sunglasses: Some code style and we're done ;-)

**Comment by georgehrke** [comment]
@weeman1337 I actually developed something very similar lately, but as an independent app instead of putting it into the server.

I have to two fix some minor issues and will push it to a repository tonight. 
Maybe we can merge our efforts together? :) 

signal-attachment-2019-01-13-213155
signal-attachment-2019-01-13-222419

**Comment by weeman1337** [comment]
@georgehrke thanks for the info. So that will be the official "where is my data" app?

That would mean this PR is obsolete?

**Comment by MorrisJobke** [comment]
> @georgehrke thanks for the info. So that will be the official "where is my data" app?
> 
> That would mean this PR is obsolete?

I would not say so. I would like to have this as it is in this PR. Then we can think on how to integrate both of them into one interface. The problem here was that @karlitschek was not aware of this PR and asked @georgehrke to do it in an app.

**Comment by karlitschek** [comment]
Yes. It's totally my fault that I wasn't aware of the duplicate work. Sorry. Not sure what the best way is to proceed. Whatever you suggest. @MorrisJobke @weeman1337 @georgehrke

**Comment by MorrisJobke** [comment]
@georgehrke @weeman1337 I would get this PR here in because it is fully working and was tested already. Also the design team had a look at it and it's approved. We then can check out how to combine or integrate it with your approach because you feature a lot more context due to the user data manifesto.

**Comment by MorrisJobke** [comment]
@skjnldsv @danielkesselberg @georgehrke @rullzer Please review.

File: lib/private/Settings/Personal/PersonalInfo.php [hunk 3]
```diff
@@ -135,12 +141,40 @@ public function getForm() {
...
 
 		return new TemplateResponse('settings', 'settings/personal/personal.info', $parameters, '');
 	}
 
+	/**
+	 * Returns the "where is your data" template params.
+	 *
+	 * @return array
+	 */
+	private function getWhereIsYourDataParams() {
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply lib/private/Settings/Personal/PersonalInfo.php]: ✓ so for NC16 I can assume PHP >= 7.1 :)

File: settings/Controller/ServerInfoSettingsController.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	 * ServerInfoSettingsController constructor.
+	 *
+	 * @param IConfig $config
+	 */
+	public function __construct($appName, IRequest $request, IConfig $config) {
+		parent::__construct($appName, $request);
+		$this->config = $config;
+	}
+
+	public function storeServerInfo(
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply settings/Controller/ServerInfoSettingsController.php]: ✓ so for NC16 I can assume PHP >= 7.1 :)

**Review by kesselb — APPROVED** [review]

File: lib/private/Settings/Personal/PersonalInfo.php [hunk 5]
```diff
@@ -38,6 +39,7 @@
 use OCP\IUserManager;
 use OCP\L10N\IFactory;
 use OCP\Settings\ISettings;
+use OCP\Encryption\IManager as EncryptionManager;
```
Comment (danielkesselberg) [inline lib/private/Settings/Personal/PersonalInfo.php]: IEncryptionManager would be more clear here.

Comment (MorrisJobke) [reply lib/private/Settings/Personal/PersonalInfo.php]: Let's do this later as well together with the email check.

**Pull request merged**